### PR TITLE
feat: block out of bounds symlinks (#9738)

### PR DESCRIPTION
**Pull request by notfromstatefarm**
Signed-off-by: notfromstatefarm <86763948+notfromstatefarm@users.noreply.github.com>

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)). 

Fixes #9738

This PR implements @crenshaw-dev's idea of traversing the repository for symlinks and blocking further processing if any are found to traverse outside of the directory, even if the final target is within it. The check can be disabled via the repo server argument `allow-oob-symlinks`.

I believe this should cover all our bases since it is implemented inside `runRepoOperation`. It will print to the console whenever any are found.

I had to change many of the repository tests because they have the root of the mock repo being the root of the Argo repository. This meant that any time it 'fetched' the repository, it was finding the test files I created for the unit tests and thereby failing as it should.

Creating new app:
image

Refreshing an app that had bad symlinks added to it:
image

```
argocd-repo-server time="2022-07-01T03:52:31Z" level=warning msg="repository contains out-of-bounds symlink. repo: https://github.com/notfromstatefarm/argocd-symlink-test.git revision: f9c92205557f4b02e63268d1964e6a0acdb83310 file: badlinksinhere/badlink"
```

### CODE REVIEW

**Comment by notfromstatefarm** [comment]
Hah. It works SO good, it broke another test because it errored out complaining about the symlinks! Gotta move my test files somewhere else!

**Comment by notfromstatefarm** [comment]
Okay, I think I fixed the repository tests properly. From what I can tell, the existing tests weren't taking into account being blocked by this new feature.

**Comment by notfromstatefarm** [comment]
Had to fight with tests for a while until I found the right approach. Turns out by implementing this, it broke most of the repository tests because they were scoped with the root of the 'app' being the root of the repository. This meant that any time it 'loaded' the repository, it was finding the test files I created for the unit tests and thereby failing.

To avoid this and still have the unit test files, I had to tweak most of the tests in repository_test.go.

**Comment by crenshaw-dev** [comment]
You gotta quit putting up PRs so fast, I can't review them all! 😆 But for real, thanks for working on this.

> Hah. It works SO good, it broke another test because it errored out complaining about the symlinks! 

lol, I'm not surprised. Hopefully you were able to sort it.

A bunch of tests use relative paths, when the code really should only accept absolute paths.

File: reposerver/repository/repository.go
Comment (crenshaw-dev) [inline reposerver/repository/repository.go:328]: How about `log.WithFields` for the chart, revision, and file?

File: util/app/path/path.go
Comment (crenshaw-dev) [inline util/app/path/path.go:47]: We have a files.IsSymlink helper method if you wanna use that.

File: util/app/path/path.go
Comment (crenshaw-dev) [inline util/app/path/path.go:36]: For the places that return an error, can you return `fmt.Errorf(...` to add a bit of context to the error message? Should make debugging easier.

File: util/app/path/path.go
Comment (crenshaw-dev) [inline util/app/path/path.go:58]: This is actually kinda funny. While it's true that technically an absolute path could point in-bounds, you'd have to know the path (which should be secret). I like it.  :-)

Comment (crenshaw-dev) [reply util/app/path/path.go:58]: Could this check move up above the `filepath.Rel` check, since it doesn't depend on that var?

File: util/app/path/path.go
Comment (crenshaw-dev) [inline util/app/path/path.go:76]: Instead of setting `found` separately, can we just embed it in the original error message? Something like this:

```go
return fmt.Errorf("encountered out of bounds symlink at %q: %w", found, outOfBoundsError)
```

Then you can use `errors.Is()` for the error check here?

File: util/app/path/path.go
Comment (crenshaw-dev) [inline util/app/path/path.go:77]: Can you add a `log.Warn`? Out-of-bounds symlinks are suspicious, so it'd be nice to give admins a way to monitor them.

**Comment by notfromstatefarm** [comment]
> You gotta quit putting up PRs so fast, I can't review them all! 😆 But for real, thanks for working on this.

Sorry! I saw a low-hanging fruit, had a slow day, and I love security. 😆 

> lol, I'm not surprised. Hopefully you were able to sort it.

Seems I've broken an E2E test.. Not entirely sure how since I'm not seeing my error. I have absolutely no idea how to dig into that yet but I'll figure it out. Is it OK if I DM you on the CNCF Slack every once in a while for assistance?

File: util/app/path/path.go
Comment (crenshaw-dev) [inline util/app/path/path.go:68]: This is tricky, because `a..b` is a valid filename.

What we've done in other places is calculate the absolute path and then making sure its prefix is the expected root path (with the trailing `/`). https://github.com/argoproj/argo-cd/blob/master/util/io/files/util.go#L72-L90

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]
This is awesome. First pass review finished.  :-)

**Comment by crenshaw-dev** [comment]
> Is it OK if I DM you on the CNCF Slack every once in a while for assistance?

Sure! Generally I don't have time for support DMs, but I can always make time for contribution DMs.  :-)

File: util/app/path/path.go
Comment (notfromstatefarm) [reply util/app/path/path.go:68]: Ah crap, you're right! I forgot about that. Gonna have to think about this one. Good catch!

Comment (crenshaw-dev) [reply util/app/path/path.go:68]: I wonder if `startswith(.. + path separator)` would be sufficient.

File: util/app/path/path.go
Comment (notfromstatefarm) [reply util/app/path/path.go:58]: It does, it's setting `found = repoPath` inside the `IsAbs` check, `repoPath` comes from `filepath.Rel`

**Comment by crenshaw-dev** [comment]
Hm. Is there an e2e test with a symlink up to the base repo dir?

```
time="2022-07-01T14:46:38Z" level=error msg="`../../dist/argocd app create test-helm-repo --repo https://localhost:9444/argo-e2e/testdata.git/helm-repo/local --dest-server https://kubernetes.default.svc/ --helm-chart helm --project default --revision 1.0.0 --dest-namespace argocd-e2e--test-helm-repo-lltho --plaintext --server 127.0.0.1:8088 --auth-token *** --insecure` failed exit status 20: time=\"2022-07-01T14:46:38Z\" level=fatal msg=\"rpc error: code = InvalidArgument desc = application spec for test-helm-repo is invalid: InvalidSpecError: Unable to generate manifests in : rpc error: code = Unknown desc = open /tmp/_argocd-repo: permission denied\"" execID=b89db
```

File: util/app/path/path.go
Comment (crenshaw-dev) [inline util/app/path/path.go:57]: `os.Readlink` might be more clever than we want. Looks like it might actually evaluate symlinks. That's a problem if we want to avoid allowing the user to infer existence/non-existence of out-of-bounds paths. I think that might be why the e2e test is failing. We're running `os.Readlink` on `../../` and getting a permissions error on `/tmp/_argocd-repo`. I gotta read the docs some more.

Comment (crenshaw-dev) [reply util/app/path/path.go:57]: I also wonder if `Readlink` evaluates the path relative to the link file or CWD: https://man7.org/linux/man-pages/man2/readlink.2.html

File: util/app/path/path.go
Comment (notfromstatefarm) [reply util/app/path/path.go:66]: If what they claim is true (Macs return absolute path), the unit tests I wrote for the good symlinks would have failed when I ran it on my Mac. It passed just fine on both my Mac and our Linux-based CI. 🤔

File: util/app/path/path.go
Comment (notfromstatefarm) [reply util/app/path/path.go:55]: It's already reporting which symlink, not the destination. Commented the code so it's more clear what's going on. 😄

File: util/app/path/path.go
Comment (notfromstatefarm) [reply util/app/path/path.go]: `os.Readlink` takes an absolute path and returns the exact contents of the link 'file' without following it. It shouldn't be resolving anything - that's what `filepath.EvalSymlinks` does and why I didn't use it.

I just tested it locally too - `os.Readlink` does not complain if the symlink points to a folder that has 000 filemode.

**Comment by notfromstatefarm** [comment]
The error must be coming from the traversal itself.. Is there some way to run this E2E test specifically so we don't have to wait 30+ minutes to try things?

**Comment by notfromstatefarm** [comment]
I cannot for the life of me get E2E to work on my M1.. When I try running it locally, I get connection refused errors for port 8081. When I try running it in the docker container, I get a permission denied for `.kube/config.lock`.

Lets add some `fmt.Errorf` and figure out which part is actually causing the E2E to fail.

**Comment by notfromstatefarm** [comment]
@crenshaw-dev Okay, I FINALLY figured out how to run the E2E test locally. The problem isn't in `os.Readlink`. It's in `filepath.Walk`.

When it's a Git repo, I walk the `gitClient.Root()` path. When it's a Helm chart, I tried walking the service rootDir and apparently that is `/tmp/_argocd-repo`. I think I need to use `chartPath` instead, but that breaks a bunch of repository tests.

**Comment by notfromstatefarm** [comment]
Ok, all tests are passing locally! 🎉 

The problem was that inside of `repository/testdata/my-chart` there is an out-of-bounds symlink used to test the checks that **values files** cannot escape. Other tests based on this same chart would pass successfully so long as it didn't try to use that values file. However, now that we have a much lower-level check, this entire chart is being rejected due to the symlink.

To solve this, I had to remove the out-of-bounds symlink from `my-chart-1` and remove the associated tests. I believe this is the correct approach since it's now impossible for symlinks to link outside of the repo/helm chart, period.

**Comment by notfromstatefarm** [comment]
Added another integration test into `repository_test.go`. Interestingly, I had to make a separate testdata2 directory for it - the other tests would fail as soon as it saw the out-of-bounds symlink in the 'helm chart root' aka testdata.

File: util/app/path/path.go
Comment (crenshaw-dev) [inline util/app/path/path.go:80]: I think `string(os.PathSeparator)+"..")` is too aggressive. `a/..bc` would match despite not being a traversal.

Comment (notfromstatefarm) [reply util/app/path/path.go:80]: Good catch again!

Come to think of it, `strings.HasPrefix(rel, ".."+string(os.PathSeparator)` should be sufficient. This is checking an already cleaned path and so anything outside of the repo should begin with `../`. Thanks to the implicit `filepath.Clean` from `filepath.Rel` It's not going to have any trickery like `./..` so we don't have to worry about those edge cases.

**Comment by crenshaw-dev** [comment]
I'm considering whether we should move the goalposts _just a little_.

What if, along with preventing the symlink target from being outside the root dir, we prevented any _part_ of the symlink path from ever leaving the root dir. This would prevent an attacker from being able to brute-force (or otherwise guess) the root dir path.

For example, consider this directory structure.

```
/tmp
  /_argocd-repo
    /{randomized dir name 1}
      secret
    /{randomized dir name 2}
      link (-> ../{randomized dir name 1}/../{randomized dir name 2})
```

I believe the current check would allow this symlink to pass, because its eventual destination does not lie outside the root dir. But the passing check allows the attacker to confirm that they've correctly guessed some external path.

Granted, all the sensitive paths should be 128 bits of cryptographically secure entropy. But this would be an extra layer of protection in case of a particularly persistent adversary or a CSPRNG flaw.

I think the solution would involve splitting the link path and then reconstructing it, checking for an out-of-bounds error at each step. It would be all string manipulation operations (no filesystem work), so should be reasonably fast.

**Comment by crenshaw-dev** [comment]
But if we want to save that work for another time, that's understandable. 😆

File: util/app/path/path.go
Comment (crenshaw-dev) [reply util/app/path/path.go]: Agreed, I think the `../` check is fine!

**Comment by notfromstatefarm** [comment]
Good thinking, but I'm not sure we need to be concerned with that. The `filepath` methods don't care whether or not parts of the paths/symlinks exist - ultimately your example of `../{randomized dir name 1}/../{randomized dir name 2}` will just be resolved to `../{randomized dir name 2}`.

**Comment by notfromstatefarm** [comment]
I had to think about this one for a minute, but you're right! Even though the ultimate target can't be outside of the repository, the intermediate directories in the link must exist or it will fail.

You're really good at thinking of these edge cases. 😄 

I also seem to have introduced a flaky test.. Charts are switching spots. Bah.

**Comment by crenshaw-dev** [comment]
> You're really good at thinking of these edge cases. 😄

Comes from pondering about 6 CVEs related to path traversal. 😆 

I'll take a look at the flaky tests in a bit. Being lazy on the weekend so far and only thinking of ways to make things more difficult rather than more easy. 😉

**Comment by notfromstatefarm** [comment]
I think this'll do the trick! It now traverses each step of the symlink to make sure it never leaves the bounds of the repo.

File: util/app/path/path.go
Comment (crenshaw-dev) [inline util/app/path/path.go:81]: `if rel == ".."` - good catch

Comment (notfromstatefarm) [reply util/app/path/path.go:81]: yeah, I quickly realized it wasn't working on the first level - that's why! Shoulddddd be all good to go now (sans flaky test) unless your magical brain can find something else 🤣

**Comment by notfromstatefarm** [comment]
Okay, flaky test should be fixed. Ran it 100x locally. I think that's everything!

File: docs/operator-manual/argocd-cmd-params-cm.yaml
Comment (crenshaw-dev) [inline docs/operator-manual/argocd-cmd-params-cm.yaml:117]: +

```
Changing this to "true" will not allow _all_ out-of-bounds symlinks. Those will still be blocked for things like values files in Helm charts. But symlinks which are not explicitly blocked by other checks will be allowed.
```

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]
I'm going to give this one last pass before smashing merge. But I'm 99.9% sure it's 100% awesome. :-) 

Would you kindly add docs/operator-manual/upgrading/2.4-2.5.md and include a note about this change? Ideally the note would explain that we're trying to prevent 1) out-of-bounds symlinks that might be missed by other checks and 2) leaking _any_ information about the presence or absence of files or directories outside the app's repo.

Also feel free to say "naw I'm busy," and I'll add the note later.  :-)

File: docs/operator-manual/upgrading/2.4-2.5.md
Comment (crenshaw-dev) [inline docs/operator-manual/upgrading/2.4-2.5.md:3]: Damn. This is well-written. Thanks!

File: reposerver/repository/repository_test.go
Comment (crenshaw-dev) [inline reposerver/repository/repository_test.go:783]: Is it possible to use `assert.ErrorIs` here?

File: util/app/path/path.go
Comment (crenshaw-dev) [inline util/app/path/path.go:76]: Is `string(os.PathSeparator)+` necessary here?

File: util/app/path/path.go
Comment (crenshaw-dev) [inline util/app/path/path.go:74]: I think you could just use filepath.SplitList here. The only difference should be that if the link is an empty string, `parts` will be `[]` instead of `[""]`.

File: util/app/path/path_test.go
Comment (crenshaw-dev) [inline util/app/path/path_test.go:57]: Could we use `ErrorIs` for these?

**Review by crenshaw-dev — APPROVED** [review]
Down to optional nitpicks. I'm "go" for merge.  :-) Going to look for someone to do a second review for good measure.

File: util/app/path/path.go
Comment (notfromstatefarm) [reply util/app/path/path.go:74]: Changing it, for some reason or another, broke bad link test 3.

File: util/app/path/path.go
Comment (notfromstatefarm) [reply util/app/path/path.go]: Nope, removed!

File: reposerver/repository/repository_test.go
Comment (notfromstatefarm) [reply reposerver/repository/repository_test.go:783]: Not that I'm aware of. At this point it's a new error and not my custom type.

Comment (crenshaw-dev) [reply reposerver/repository/repository_test.go:783]: New as in it's wrapped? If you use `fmt.Errorf("... %w", wrapperErr)`, then `errors.As` will unwrap until it finds the error.

But don't sweat it.  :-)

Comment (notfromstatefarm) [reply reposerver/repository/repository_test.go:783]: Oooooh. Learned me some more Go today!

File: reposerver/repository/repository.go
Comment (pasha-codefresh) [inline reposerver/repository/repository.go:323]: just cosmetic change , i think it would be nice move this block to some dedicate function, so you will not need duplicate it fully in bottom

Comment (crenshaw-dev) [reply reposerver/repository/repository.go:323]: Gonna go ahead and merge this, and @notfromstatefarm can follow up with another PR if this can be DRY'ed out.

**Pull request merged**
